agent/containermetadata/manager.go

```diff
@@ -0,0 +1,166 @@
+// statuses should amended/appended accordingly.
+type MetadataStatus string
+
+// Manager is an interface that allows us to abstract away the metadata
+// operations
+type Manager interface {
+	SetContainerInstanceARN(string)
+	Create(*docker.Config, *docker.HostConfig, string, string) error
+	Update(string, string, string) error
+	Clean(string) error
+}
+
+// metadataManager implements the Manager interface
+type metadataManager struct {
+	// client is the Docker API Client that the metadata manager uses. It defaults
+	// to 1.21 on Linux and 1.24 on Windows
+	client DockerMetadataClient
+	// cluster is the cluster where this agent is run
+	cluster string
+	// dataDir is the directory where the metadata is being written. For Linux
+	// this is a container directory
+	dataDir string
+	// dataDirOnHost is the directory from which dataDir is mounted for Linux
+	// version of the agent
+	dataDirOnHost string
+	// containerInstanceARN is the Container Instance ARN registered for this agent
+	containerInstanceARN string
+	// osWrap is a wrapper for 'os' package operations
+	osWrap oswrapper.OS
+	// ioutilWrap is a wrapper for 'ioutil' package operations
+	ioutilWrap ioutilwrapper.IOUtil
+}
+
+// NewManager creates a metadataManager for a given DockerTaskEngine settings.
+func NewManager(client DockerMetadataClient, cfg *config.Config) Manager {
+	return &metadataManager{
+		client:        client,
+		cluster:       cfg.Cluster,
+		dataDir:       cfg.DataDir,
+		dataDirOnHost: cfg.DataDirOnHost,
+		osWrap:        oswrapper.NewOS(),
+		ioutilWrap:    ioutilwrapper.NewIOUtil(),
+	}
+}
+
+// SetContainerInstanceARN sets the metadataManager's ContainerInstanceArn which is not available
+// at its creation as this information is not present immediately at the agent's startup
+func (manager *metadataManager) SetContainerInstanceARN(containerInstanceARN string) {
+	manager.containerInstanceARN = containerInstanceARN
+}
+
+// Create creates the metadata file and adds the metadata directory to
+// the container's mounted host volumes
+// Pointer hostConfig is modified directly so there is risk of concurrency errors.
+func (manager *metadataManager) Create(config *docker.Config, hostConfig *docker.HostConfig, taskARN string, containerName string) error {
+	// Create task and container directories if they do not yet exist
+	metadataDirectoryPath, err := getMetadataFilePath(taskARN, containerName, manager.dataDir)
+	// Stop metadata creation if path is malformed for any reason
+	if err != nil {
+		return fmt.Errorf("container metadata create for task %s container %s: %v", taskARN, containerName, err)
+	}
+
+	err = manager.osWrap.MkdirAll(metadataDirectoryPath, os.ModePerm)
+	if err != nil {
+		return fmt.Errorf("creating metadata directory for task %s: %v", taskARN, err)
+	}
+
+	// Acquire the metadata then write it in JSON format to the file
+	metadata := manager.parseMetadataAtContainerCreate(taskARN, containerName)
```

Comment:
The logic of parsing, marshaling and writing the metadata file is duplicated between this and the `Update` method. Please extract that to a method of its own.

Reply:
fixed.